crates/qasm2/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#[macro_use]
+extern crate lazy_static;
+
+use pyo3::prelude::*;
+use pyo3::Python;
+
+mod bytecode;
+mod error;
+mod expr;
+mod lex;
+mod parse;
+
+/// Information about a custom instruction that Python space is able to construct to pass down to
+/// us.
+#[pyclass]
+#[derive(Clone)]
+pub struct CustomInstruction {
+    pub name: String,
+    pub num_params: usize,
+    pub num_qubits: usize,
+    pub builtin: bool,
+}
+
+#[pymethods]
+impl CustomInstruction {
+    #[new]
+    fn __new__(name: String, num_params: usize, num_qubits: usize, builtin: bool) -> Self {
+        Self {
+            name,
+            num_params,
+            num_qubits,
+            builtin,
+        }
+    }
+}
+
+/// Information about a custom classical function that should be defined in mathematical
+/// expressions.
+///
+/// The given `callable` must be a Python function that takes `num_params` floats, and returns a
+/// float.  The `name` is the identifier that refers to it in the OpenQASM 2 program.  This cannot
+/// clash with any defined gates.
+#[pyclass(text_signature = "(name, num_params, callable, /)")]
+#[derive(Clone)]
+pub struct CustomClassical {
+    pub name: String,
+    pub num_params: usize,
+    pub callable: PyObject,
+}
+
+#[pymethods]
+impl CustomClassical {
+    #[new]
+    fn __new__(name: String, num_params: usize, callable: PyObject) -> Self {
+        Self {
+            name,
+            num_params,
+            callable,
+        }
+    }
+}
+
+/// Create a bytecode iterable from a string containing an OpenQASM 2 program.  The iterable will
+/// lex and parse the source lazily; evaluating OpenQASM 2 statements as required, without loading
+/// the entire token and parse tree into memory at once.
+#[pyfunction]
+fn bytecode_from_string(
+    _py: Python<'_>,
```

Comment:
Should be fine to leave this off, e.g. the first example in https://pyo3.rs/v0.18.1/function#python-functions

Reply:
Thanks, done in 203c7bb9e.

It might have been a historical artifact - my original implementation had templated structs for the two different bytecode iterators, and they may have interacted with the GIL on creation.  I wasn't (still am not entirely) quite sure what I was doing with file access in Rust - a part of what I wrote this to learn!